hoomd/simulation.py

```diff
@@ -172,27 +161,96 @@ def always_compute_pressure(self, value):
-        #     nl.update_exclusions_defaults()
-
-        # detect 0 hours remaining properly
-        self._cpp_sys.run(int(tsteps), 0, None, 0, 0)
+        self._cpp_sys.run(int(steps))
+
+    def write_debug_data(self, filename):
+        """Write debug data to a JSON file.
+
+        Args:
+            filename (str): Name of file to write.
+
+        The debug data file contains useful information for others to help you
+        troubleshoot issues.
+
+        Note:
+            The file format and particular data written to this file may change
+            from version to version.
+
+        Warning:
+            The specified file name will be overwritten.
+        """
+        debug_data = {}
+        debug_data['hoomd_module'] = str(hoomd)
+        debug_data['version'] = dict(
+            compile_date=hoomd.version.compile_date,
+            compile_flags=hoomd.version.compile_flags,
+            cxx_compiler=hoomd.version.cxx_compiler,
+            enable_gpu=hoomd.version.enable_gpu,
+            enable_mpi=hoomd.version.enable_mpi,
+            enable_tbb=hoomd.version.enable_tbb,
+            git_branch=hoomd.version.git_branch,
+            git_sha1=hoomd.version.git_sha1,
+            gpu_api_version=hoomd.version.gpu_api_version,
+            gpu_platform=hoomd.version.gpu_platform,
+            install_dir=hoomd.version.install_dir,
+            source_dir=hoomd.version.source_dir)
+
+        reasons = hoomd.device.GPU.get_unavailable_device_reasons()
+
+        debug_data['device'] = dict(
+            msg_file=self.device.msg_file,
+            notice_level=self.device.notice_level,
+            devices=self.device.devices,
+            num_cpu_threads=self.device.num_cpu_threads,
+            gpu_available_devices=hoomd.device.GPU.get_available_devices(),
+            gpu_unavailable_device_reasons=reasons)
+
+        debug_data['communicator'] = dict(
+            num_ranks=self.device.communicator.num_ranks)
+
+        # TODO: Domain decomposition
+
+        if self.state is not None:
+            debug_data['state'] = dict(
+                types=self.state.types,
+                N_particles=self.state.N_particles,
+                N_bonds=self.state.N_bonds,
+                N_angles=self.state.N_angles,
+                N_impropers=self.state.N_impropers,
+                N_special_pairs=self.state.N_special_pairs,
+                N_dihedrals=self.state.N_dihedrals,
+                box=repr(self.state.box))
+
+        # save all loggable quantities from operations and their child computes
+        logger = hoomd.logging.Logger(only_default=False)
+        logger += self
+
+        # children may appear several times, identify them uniquely
+        children = list()
+        for op in self.operations:
+            if op is not None:
+                logger.add(op)
+
+                for child in op._children:
+                    if child not in children:
+                        children.append(child)
```

Comment:
It does not appear multiple times, but I do not know why.

Reply:
When adding objects to the logger we check if that exact object has been logged in that namespace before (i.e. `obj1 is obj2`). Since the neighbor list will always return the same namespace, we see it is already in use and that it is in use by the same object and so perform a no-op.